### ISSUE TITLE
Running the model in the browser

### ISSUE THREAD

**Issue by gabrieldemarmiesse** [kyutai-labs/pocket-tts#1]
It would be awesome to have a version running purely in the browser. This TTS is quite fast on CPU already, so we don't actually need to access the gpu from the browser. If someone is motivated to re-implement the model or export it from the pytorch version to run it in the browser, we'd be happy to host the resulting html page!Feel free to make a pull request to this repository and add the implementation in a directory called `in_the_browser/`.

 Choose whatever tech and whatever path you want to get there. 
As long as it works™

**Comment by xenova** [comment]
Hey! 👋  I'm interested in tackling this :) Have there been any efforts to convert the model to ONNX, for example?

**Comment by gabrieldemarmiesse** [comment]
We don't have any code to convert to ONNX, feel free to try. Other approaches like using tensorflow.js are possible too if exporting to ONNX is too annoying :)

**Comment by gabrieldemarmiesse** [comment]
Hello, let's clarify here that you don't need permission to work on any issue. You also don't need to say that you're working on it. Just make pull requests and we'll be very happy to review them! :) We don't assign people to issues. See the [rules about issues](https://github.com/kyutai-labs/pocket-tts/blob/main/CONTRIBUTING.md#issues-etiquette) in the contributing guide.

I'll delete the previous messages so that future readers don't think that the issue is assigned to somebody.

**Comment by ekzhang** [comment]
Hmm I see @xenova you've commented but down to collaborate on this with you, whether through ONNX or a manual implementation in jax-js. I feel this may be very nice model to run in the browser locally; it sounds much more natural than the Web Audio API at least in the demo and the voice cloning is really cool as well

**Comment by X-Ryl669** [comment]
I think using ONNX (that is, removing python code between each step of the processing and only using pytorch here) is the best path forward, since converting a pytorch processing pipeline to ONNX is very easy and converting ONNX to web assembly is too.

**Comment by ekzhang** [comment]
Took some general notes about the model architecture [here](https://github.com/ekzhang/jax-js/issues/59) in case it's useful to anyone else attempting this haha

**Comment by xenova** [comment]
> I see @xenova you've commented but down to collaborate on this with you, whether through ONNX or a manual implementation in jax-js

Would love to! :) I think an ONNX conversion/implementation should be possible. To start, we'd aim to break the model up into logical parts. After creating wrapper classes for each, we can utilize torch.onnx.export to handle the conversion.

**Comment by ekzhang** [comment]
As a quick progress update, I've gotten the model running locally in the web browser with inference! It runs close to real-time with WebGPU on macOS.

Spent most of today implementing/debugging the model architecture in code — there's a SEANet for convolutional upsampling with residual blocks, Mimi neural audio codec (with kv cache, 2 seconds audio context for local attention), and flow network backbone.

https://github.com/user-attachments/assets/0c08decf-10bc-4696-a560-c6b2c49910a8

Here is a runnable preview of it: https://jax-js-git-pocket-tts-ekzhang.vercel.app/tts — for some reason it even sounds better than the serverless function version?

There is a lot of obvious low-hanging optimizations, e.g., it's running without any kv caches right now, and it needs to compile a bunch of kernels on first run due to a lack of padding. Anyway I'll implement this soon.

**Comment by bouba1ml** [comment]
I vibe-coded myself one, and everything is working, including voice cloning ( https://github.com/bouba1ml/myWebPocketTTS/tree/main)
Hope everything was committed and correctly (not used to Github yet)

<img width="764" height="861" alt="Image" src="https://github.com/user-attachments/assets/2decd305-b1dc-4857-9154-207d801d7535" />

**Comment by ekzhang** [comment]
@gabrieldemarmiesse — I have a working version currently hosted on the main website for `jax-js` here: https://jax-js.com/kyutai-pocket-tts including a link to the source code - it's actually at `/tts`, but I wanted to give this model a more permanent / future-proof link so `/kyutai-pocket-tts` redirects to `/tts` for now.

Does this work for you?

I could try and make it more standalone and make a PR to this repo, but it might be easier to keep it on the jax-js repository for now as it will benefit from some perf improvements that I'm still working on (e.g., faster attention kernel), and I can update it if the library changes.

**Comment by vvolhejn** [comment]
@ekzhang I already messaged you on X but I'm only seeing your comment now here.

It looks like there are three implementations at the moment:
- @ekzhang's jax-js implementation [here](https://jax-js.com/kyutai-pocket-tts)
- @babybirdprd's Rust/Candle implementation: #43 
- @kevinahm's ONNX export [here](https://github.com/kyutai-labs/pocket-tts/issues/2#issuecomment-3766066669)

For now I'll link to all of these in the README, eventually it'd be great to have one "official" web implementation integrated in this repo. I'd say the considerations are speed, ease of integration, and maintainability. It's likely that we'll figure this out together with Gabriel when he's back on February 2.

If you would *not* like us to integrate your implementation in this repo, let us know so we can plan for that :)

**Comment by ekzhang** [comment]
Nice, sounds good — and wow the Wasm implementations are indeed very impressive

**Comment by gabrieldemarmiesse** [comment]
@ekzhang @babybirdprd @KevinAHM  I tried the three implementations and here are my findings:
* Jax-js: not real-time on my cpu (intel core ultra 7 165H), but extremely long prompts are working. Since jax is fully static shapes, maybe it's just a matter of tweaking the attention size?
* WASM with rust: There is a significant delay of around 15s before giving the audio (the audio is 5s long). I guess something is under-optimized? And maybe streaming is not enabled? Very long prompts are working though, which is very nice.
* ONNX in the browser: works in real time!!! But I couldn't use very long prompts. After 4 big sentences it stops (it's not just the limit of 500  on the number of characters, the audio stops before reading all the text). I guess it's missing the splitting/chunking we do in the python version?

Thanks all for the work on this! It seems we're not quite ready to replace our demo on the website by an in-browser implementation but I hope we can do so soon!

**Comment by KevinAHM** [comment]
@gabrieldemarmiesse 
I've fixed the ONNX browser implementation to use the proper chunking, similar to the Python implementation. Now the space can generate an unlimited length of text.

**Comment by gabrieldemarmiesse** [comment]
@KevinAHM  I just tried it, it works great! :)

**Comment by ekzhang** [comment]
@gabrieldemarmiesse The jax-js implementation is using WebGPU, it's likely not real-time on the Windows machines but runs on macOS M1 and on iPhone with no issues.

We don't have a jax-js CPU version yet as the Wasm backend doesn't implement SIMD or mutltithreading with SharedArrayBuffer.

Fwiw the splitting/chunking is also missing from jax-js one, it's fine though since the GPU can plumb through all of the attention FLOPs haha

**Comment by gabrieldemarmiesse** [comment]
It is surprising though because my laptop has a fairly recent nvidia gpu. Maybe it's not using it for some reason?

**Comment by babybirdprd** [comment]
@gabrieldemarmiesse Was this delay on only the first generation, or rather was rhe delay worse on the first generation than any following that? I'm guessing there is something under optimized, I didnt spend much time on wasm, most of my effort was on using it from the cli. I will look into it, and update here.

**Comment by ekzhang** [comment]
Oh interesting Gabriel. Yeah not sure then, in any case I have a bit to work on! jax-js has only been out for a month and running pocket-tts on it has been a great test of the kernels and library, I'll probably continue improving things

**Comment by gabrieldemarmiesse** [comment]
Just FYI, for our web demo in the browser, we'll go for a cpu-only implementation, as we want the demo to work well anywhere, even on potato laptops and phones. But obviously feel free to do whatever you want in your own implementations.

Concerning the rust version, the delay of 15 seconds at the start was present every time, no only the first time.

**Comment by babybirdprd** [comment]
> Just FYI, for our web demo in the browser, we'll go for a cpu-only implementation, as we want the demo to work well anywhere, even on potato laptops and phones. But obviously feel free to do whatever you want in your own implementations.
> 
> Concerning the rust version, the delay of 15 seconds at the start was present every time, no only the first time.

I'm working on this now, what would you say is an acceptable latency?

**Comment by gabrieldemarmiesse** [comment]
I would say sub-600ms would be acceptable. If that's not possible, no worries, I don't want to impose work onto you. We could go for the ONNX version if the performance of candle in the browser is not great.

**Comment by babybirdprd** [comment]
> I would say sub-600ms would be acceptable. If that's not possible, no worries, I don't want to impose work onto you. We could go for the ONNX version if the performance of candle in the browser is not great.

I think that should be possible. I might just have to apply some of the changes from the web app, to the wasm one. I don't mind spending some time on it, having it run in the browser is a nice feature to have. Maybe later today I can look into it further, I was able to reduce the latency quite a bit, but havent measured it yet.

**Comment by csukuangfj** [comment]
[sherpa-onnx](https://github.com/k2-fsa/sherpa-onnx) also supports running pocket tts with WebAssembly in your browser.
See https://huggingface.co/spaces/k2-fsa/web-assembly-en-tts-pocket

Also, `sherpa-onnx` provides 12 programming language APIs for pocket tts. See also
https://k2-fsa.github.io/sherpa/onnx/tts/pocket.html

<img width="1826" height="1812" alt="Image" src="https://github.com/user-attachments/assets/61719751-08aa-4275-b079-7171636d11bc" />

**Comment by gabrieldemarmiesse** [comment]
We now have many in browser implementations of Pocket TTS. Really a huge thank you to the community for all this awesome work! If you have other implementations you'd like to share with the world, don't hesitate to make a PR to add it to the readme!
